src/engine/strat_engine/range_alloc.rs

```diff
@@ -0,0 +1,232 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+use std::cmp::min;
+use std::u64::MAX;
+
+use devicemapper::Sectors;
+use ranged_set::RangedSet;
+
+#[derive(Debug)]
+pub struct RangeAllocator {
+    limit: Sectors,
+    used: RangedSet<u64>,
+}
+
+impl RangeAllocator {
+    /// Create a new RangeAllocator with all entries in the range unused.
+    pub fn new(limit: Sectors) -> RangeAllocator {
+        assert!(*limit < MAX, "limit must be less than 2^64");
+        RangeAllocator {
+            limit: limit,
+            used: RangedSet::new(),
+        }
+    }
+
+    /// Create a new RangeAllocator with the specified (offset,
+    /// length) ranges marked as used.
+    pub fn new_with_used(limit: Sectors, initial_used: &[(Sectors, Sectors)]) -> RangeAllocator {
+        let mut allocator = RangeAllocator::new(limit);
+        allocator.insert_ranges(initial_used);
+        allocator
+    }
+
+    fn check_for_overflow(&self, off: Sectors, len: Sectors) {
+        assert_ne!(off.checked_add(*len), None);
+        assert!(off + len <= self.limit, "off+len greater than range limit");
+    }
+
+    fn insert_ranges(&mut self, ranges: &[(Sectors, Sectors)]) -> () {
```

Comment:
This function could use a header comment to explain it better. I think panicking when inserted value is already there is correct thing to do, but it needs justification.